### ISSUE TITLE
Intent queue is constantly overflowing 

### ISSUE THREAD

**Issue by pchojnacki** [hashicorp/consul#1062]
In our setup that was recently upgraded to 0.5.2 we currently have two different Consul DCs/clusters configured in one physical Datacenter.
One is 'production' only hosts that have better monitoring and are meant to serve production traffic. The second is meant for development only.

Before migrating to 0.5.2 we also experienced problems described here:
https://github.com/hashicorp/consul/issues/843

Now the main problem is that we receive tons of logs like these from the dev cluster:
```
    2015/06/26 13:28:39 [WARN] serf: Intent queue depth: 4256
    2015/06/26 13:28:39 [WARN] serf: Intent queue depth (4256) exceeds limit (4096), dropping messages!
```
Essentially every agent is broadcasting log lines similar to almost every second.
While cluster seems fine we are worried that some performance might be degraded. 

To debug this issue I've run one of the agents in debug mode. Where I get tons of messageJoinType interspersed with warnings about queue depth. The nodes mentioned in messageJoinType are *always* from different cluster.

E.g.
```
   2015/06/26 13:27:08 [DEBUG] serf: messageJoinType:  XXXXXXse3
    2015/06/26 13:27:08 [DEBUG] serf: messageJoinType:  XXXXXX-s2
    2015/06/26 13:27:08 [DEBUG] serf: messageJoinType:  XXXXXX-s2
    2015/06/26 13:27:08 [DEBUG] serf: messageJoinType:  XXXXXXsc3
    2015/06/26 13:27:08 [WARN] serf: Intent queue depth: 2313
    2015/06/26 13:27:08 [DEBUG] serf: messageJoinType:  XXXXXXrev
    2015/06/26 13:27:08 [DEBUG] serf: messageJoinType:  XXXXXXsf1
    2015/06/26 13:27:08 [DEBUG] serf: messageJoinType:  XXXXXXsc5
    2015/06/26 13:27:08 [DEBUG] serf: messageJoinType:  XXXXXX-de
```
and
```

    2015/06/26 13:28:39 [DEBUG] serf: messageJoinType:  XXXXXX-s1
    2015/06/26 13:28:39 [DEBUG] serf: messageJoinType:  XXXXXXs10
    2015/06/26 13:28:39 [DEBUG] serf: messageJoinType:  XXXXXXsc3
    2015/06/26 13:28:39 [WARN] serf: Intent queue depth: 4256
    2015/06/26 13:28:39 [WARN] serf: Intent queue depth (4256) exceeds limit (4096), dropping messages!
    2015/06/26 13:28:39 [DEBUG] serf: messageJoinType:  XXXXXX-s7
    2015/06/26 13:28:39 [DEBUG] serf: messageJoinType:  XXXXXXa12
    2015/06/26 13:28:39 [DEBUG] serf: messageJoinType:  XXXXXXs13
```

I've also tried latest master branch Consul  and the result is the same.
I haven't tried full cluster restarts yet with cleaning up data files etc. And preferably I'd like to avoid that.

**Comment by jaredcurtis** [comment]
We're also having the same issue.

**Comment by slackpad** [comment]
Do you know if these started happening when you crossed a certain number of nodes, or right away when you did the upgrade? Can you share approximately how many nodes you have running together?

**Comment by jaredcurtis** [comment]
In our case we haven't done a full upgrade, the majority of our nodes run 0.5.2 but some are still running 0.4.1. I'm not sure when the error started but I do know that we've been cross joined for some time. Today I noticed the error in the logs a found this ticket and then the reference to #843. At this point I believe that I've removed all of the cross joined members.

We only had members from DC2 joined to DC1, 417 nodes. DC2 is showing the error message, while DC1 appears to be operating normally.

**Comment by slackpad** [comment]
Hmm - I know this was fixed in 0.5.1 - is there any pattern that implicates the older nodes as being the ones causing the problem?

**Comment by jaredcurtis** [comment]
Not sure what I would look for to identify which nodes are causing a problem.

In DC2 we have 12 nodes running 0.4.1 and 413 nodes running 0.5.2

**Comment by slackpad** [comment]
I was thinking similar @pchojnacki where he noticed that all the the messageJoin logs were from the other cluster - was curious if you saw only nodes with the older version of Consul.

**Comment by jaredcurtis** [comment]
I've now cleared up our cross issue and upgraded every node to 0.5.2. We're still seeing the problem. We're seeing lots of nodes joining and then timing out. I'm verifying all of our firewall rules now.

**Comment by pchojnacki** [comment]
Hi, sorry for such late update. But since creating this ticket I was able to clear this issue that affected our setup. 
So after digging into consul source code I decided that the easiest approach would be to momentarily stop some of the affected nodes wait for the issue to clear up and then restart them.

tl; dr I'm pretty sure the issue is too small ```recentIntent``` buffer size that is checked in this [method in Serf] (https://github.com/hashicorp/serf/blob/master/serf/serf.go#L1481) 

IIRC there were around 220 distinct messageJoinType messages passed around in the cluster and the default length of recentIntent buffer is 128 so given enough hosts communicating with each other it seems possible that it would cause for such messages to be rebroadcasted indefinitely. 

I thought about increasing the recentIntent buffer length and deploying modified consul to affected hosts but since it was a dev cluster so I could just stop them for the required time.

PS
@slackpad all nodes were running consul 0.5.2 at that time

**Comment by armon** [comment]
@pchojnacki This theory makes a lot of sense. We should probably do a combination of increasing the queue size, but also gating just dropping intents if they are "too old" to avoid the infinite broadcasts.

**Comment by jaredcurtis** [comment]
We're still seeing this issue, does anyone have any next steps? Should we move forward with a custom build? Any thing that we can look for to verify that this isn't a problem on our side?

**Comment by armon** [comment]
@jaredcurtis Probably the best short term fix is a custom build that sets the intent buffer to a much larger size. That is a bit of a hack, and we need to think through a longer term solution, but there isn't an obvious answer.

**Comment by slackpad** [comment]
Notes on a possible solution from @armon: move the intent data to be stored with the node records instead of a shared pool.
